adapters/mqtt-vertx-base/src/test/java/org/eclipse/hono/adapter/mqtt/AbstractVertxBasedMqttProtocolAdapterTest.java

```diff
@@ -1000,6 +1004,32 @@ public void testUnauthenticatedConnectionMetrics() {
         verify(metrics).decrementUnauthenticatedConnections();
     }
 
+    /**
+     * Verifies that the connection is rejected due to the limit exceeded.
+     */
+    @Test
+    public void testConnectionsLimitExceeded() {
+
+        // GIVEN an adapter requiring devices to authenticate endpoint
+        final MqttServer server = getMqttServer(false);
+        config.setAuthenticationRequired(true);
+        final AbstractVertxBasedMqttProtocolAdapter<MqttProtocolAdapterProperties> adapter = getAdapter(server);
+        forceClientMocksToConnected();
+
+        // WHEN a device tries to establish a connection
+        when(authHandler.authenticateDevice(any(MqttContext.class)))
+                .thenReturn(Future.succeededFuture(new DeviceUser("DEFAULT_TENANT", "4711")));
+        when(resourceLimitChecks.isConnectionLimitExceeded(any(TenantObject.class)))
+                .thenReturn(Future.failedFuture("Connections limit exceeded"));
+        final MqttEndpoint endpoint = getMqttEndpointAuthenticated();
+        adapter.handleEndpointConnection(endpoint);
+
+        // THEN the adapter has tried to authenticate the device
+        verify(authHandler).authenticateDevice(any(MqttContext.class));
+        // THEN the connection request is rejected
+        verify(endpoint).reject(MqttConnectReturnCode.CONNECTION_REFUSED_NOT_AUTHORIZED);
```

Comment:
we now have two different outcomes for a similar problem:
1. if the maximum number of connections for the adapter instance is exceeded (as indicated by `getConnectionLimitManager().isLimitExceeded()`, we return a `MqttConnectReturnCode.CONNECTION_REFUSED_SERVER_UNAVAILABLE`
2. if the overall maximum number of connections for the tenant is exceeded (as indicated by `ResourceLimitChecks.isConnectionLimitExceeded()`, we return a `MqttConnectReturnCode.CONNECTION_REFUSED_NOT_AUTHORIZED`

that doesn't seem right.

Also, IMHO it is a little confusing and cumbersome to have two different places where connection limits need to be configured/enforced. Maybe we can merge the existing `ConnectionLimitManager` with the `ResourceLimitChecks` so that we have only one place for doing so? But we can also do that in a second step ...

Reply:
@sophokles73 Absolutely. I also noticed that inside the `service-base` module with package as `limiting` containing `ConnectionLimitManager` and another named `plan`containing `ResourceLimitChecks` confusing. I agree that its better to merge it. I would suggest a separate PR for it.

Reply:
can we then please use the same return code?

Reply:
@sophokles73 Sure. IMHO `MqttConnectReturnCode.CONNECTION_REFUSED_NOT_AUTHORIZED` makes more sense for this scenario than the `MqttConnectReturnCode.CONNECTION_REFUSED_SERVER_UNAVAILABLE`.  WDYT?

Reply:
Is there also a reason for your assessment?

Reply:
@sophokles73 As per the spec, the description for `CONNECTION_REFUSED_SERVER_UNAVAILABLE` is _The Network Connection has been made but the MQTT service is unavailable_. FMPOV  the mqtt service is available but not authorised to connect as the limit is exceeded. so I find the `CONNECTION_REFUSED_NOT_AUTHORIZED` more suited.

Reply:
@b-abel @dejanb @ctron 
what do you guys think about this? I tend to agree with @kaniyan that `CONNECTION_REFUSED_NOT_AUTHORIZED` is a better fit here.

Reply:
Isn't there a real difference between those two scenarios? One is about the tenant-specific quota and -- as @kaniyan points out -- a authorization issue. The other one is a technical issue, where the server is not able to process more load. So having this difference reflected in the return codes IMO makes sense. 
And there are different strategies possible for the client to react. In one case this could be monitored and eventually lead to an upgrade of the quota, in the other case a simple retry might be the right thing to do, because other instances of the protocol adapter will hopefully accept the connection.

Reply:
> And there are different strategies possible for the client to react. In one case this could be monitored and eventually lead to an upgrade of the quota, in the other case a simple retry might be the right thing to do, because other instances of the protocol adapter will hopefully accept the connection.

Simply retrying will often also work in the former case (where the quota is reached) because other devices of the tenant may already have disconnected in the meantime ...

Reply:
> > And there are different strategies possible for the client to react. In one case this could be monitored and eventually lead to an upgrade of the quota, in the other case a simple retry might be the right thing to do, because other instances of the protocol adapter will hopefully accept the connection.
> 
> Simply retrying will often also work in the former case (where the quota is reached) because other devices of the tenant may already have disconnected in the meantime ...

That was an easy target, since "retry" is the generic solution in MQTT for nearly every problem :laughing: 
But seriously, for me there is one situation where, if it persists, the user of the service (tenant) needs to act and another one, that should be solved by the operator of the service, if it happens too often. This is one of very few points where the MQTT spec has something like return codes and I don't see why we should not use them.

Reply:
I like the initial definition of:

> 1. if the maximum number of connections for the adapter instance is exceeded (as indicated by getConnectionLimitManager().isLimitExceeded(), we return a MqttConnectReturnCode.CONNECTION_REFUSED_SERVER_UNAVAILABLE
> 2. if the overall maximum number of connections for the tenant is exceeded (as indicated by ResourceLimitChecks.isConnectionLimitExceeded(), we return a MqttConnectReturnCode.CONNECTION_REFUSED_NOT_AUTHORIZED

Reply:
I tend to agree as well. The current definition is clearer from the tenant/operator points of view.

Reply:
@ctron @dejanb 
My understanding is that both of you think that returning different codes for the two different cases is the right thing to do. Is that correct?

Reply:
Yes

Reply:
Ok, then we'll have two return codes indicating the similar, but different, cases. We only need to make sure that we document this properly in the MQTT adapter user guide.